### PR TITLE
[QNN] Multiplication operator

### PR DESCRIPTION
**Pull request by tristan-arm**
Implementation of the qnn multiplication operator following the same pattern as the addition operator.

### CODE REVIEW

File: src/relay/qnn/op/mul.cc [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+namespace qnn {
+
+/*
+ * \brief Canonicalizes the QNN mul op.
+ * \param attrs The QNN concatenate attrs.
+ * \param new_args The new mutated args to the call node.
+ * \param arg_types The types of input and output.
+ * \return The sequence of Relay ops for mul op.
+ */
+Expr QnnMulCanonicalize(const Attrs& attrs, const Array<Expr>& new_args,
```
[the remark this answers is not in the document]
Comment (tristan-arm) [reply src/relay/qnn/op/mul.cc]: > I think that there is a mistake in the lowering. Can you please provide description of how the lowering instructions are decided?

Same as the addition operator - only difference is that the requantized tensors are multiplied to provide the output. Let me know which part of the instructions you see an issue with and I'll take a closer look - Thanks

**Comment by tqchen** [comment]
@tristan-arm @anijain2305 please followup on this thread :)

**Comment by anijain2305** [comment]
@tristan-arm Sorry for the huge delay. This is what the lowering should look like. Let me know what do you think about this.

~~~
Quantized mul operation

    scale_c * (Q_c - zp_c) = (scale_a * (Q_a - zp_a))   *   (scale_b * (Q_b - zp_b))

    scale_c * (Q_c - zp_c) = (scale_a * scale_b)   *   [  (Q_a - zp_a) *  (Q_b - zp_b) ]


Going from right to left on the RHS
1) Shifted_input_A = Subtract Q_a - zp_a (Precision int32)
2) Shifted_input_B = Subtract Q_b - zp_b (Precision int32)
3) Int_tensors_muliplication = Shifted_input_A * Shifted_input_B


Our equation looks like this

    scale_c * (Q_c - zp_c) = (scale_a * scale_b)   *    Int_tensors_muliplication

RHS can be thought of quantized tensor with (scale, zero_point) = (scale_a * scale_b, 0)
So, we can requantize.

4) Requantize the  Int_tensors_muliplication with (input_scale, input_zero_point) = (scale_a * scale_b, 0)
    And (output_scale, output_zero_point) = (scale_c, zp_c)
~~~

**Comment by tqchen** [comment]
ping @tristan-arm please follow up :)

**Comment by u99127** [comment]
@tqchen - tristan is on holidays , the other PR by @ekalda in #4141 should supersede this work and I think this PR should be abandoned.

**Comment by tqchen** [comment]
thanks @u99127 @tristan-arm

**Pull request closed without merging**